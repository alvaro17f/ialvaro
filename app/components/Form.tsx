"use client";
import "animate.css";
import { FormEvent, useState } from "react";
import Spinner from "@/app/components/Spinner";

export default function Form() {
	const [isSubmitting, setIsSubmitting] = useState(false);
	const [isSubmitted, setIsSubmitted] = useState(false);

	const onSubmit = (e: FormEvent<HTMLFormElement>): void => {
		e.preventDefault();
		setIsSubmitting(true);

		setInterval(() => {
			setIsSubmitting(false);
			setIsSubmitted(true);
		}, 2000);
	};
	return (
		<>
			{!isSubmitted ? (
				<section
					className="grid grid-cols-1 md:grid-cols-[1fr_2fr] place-items-center p-5 mb-5 rounded-lg text-azama-dark bg-azama-warning animate__animated animate__fadeInLeft animate__delay-0.7s"
					aria-label="form"
				>
					<div className="mb-5 md:mb-0 md:text-center">
						<h2 className="text-3xl">CONTACT ME!</h2>
						<p>It's free</p>
					</div>
					<form
						onSubmit={onSubmit}
						className="grid gap-5 p-2 place-items-center md:w-full"
					>
						<input
							className="bg-azama-base placeholder:text-azama-danger text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5 shadow-md rounded-xl shadow-azama-base"
							type="text"
							name="name"
							placeholder="Name *"
							required
						/>
						<input
							className="bg-azama-base placeholder:text-azama-danger text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5 shadow-md rounded-xl shadow-azama-base"
							type="email"
							name="email"
							placeholder="Email *"
							required
						/>
						<input
							className="bg-azama-base placeholder:text-azama-muted text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5 shadow-md rounded-xl shadow-azama-base"
							type="text"
							name="phone"
							placeholder="Phone"
						/>
						<textarea
							className="bg-azama-base text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-8 shadow-md rounded-xl shadow-azama-base placeholder:text-azama-danger"
							placeholder="Your message goes here... *"
							required
						/>
						<button
							className={`md:w-40  p-4 rounded-full shadow-md shadow-azama-danger bg-azama-base bg-opacity-100 active:bg-opacity-50 text-5xl ${
								isSubmitting && "animate__animated animate__rubberBand"
							}`}
							type="submit"
						>
							{isSubmitting ? <Spinner /> : "ðŸ”¥"}
						</button>
					</form>
				</section>
			) : (
				<section
					className="grid grid-cols-1 place-items-center p-5 mb-5 rounded-lg text-azama-dark bg-azama-warning animate__animated animate__fadeInLeft animate__delay-0.7s"
					aria-label="form"
				>
					<h3 className="text-3xl">
						Thank you!{" "}
						<span className="[text-shadow:_0_3px_0_rgb(35_33_54_/_50%)]">
							ðŸ˜€
						</span>
					</h3>
					<p>
						Your message has been received. Please check your email for
						confirmation.
					</p>
				</section>
			)}
		</>
	);
}
