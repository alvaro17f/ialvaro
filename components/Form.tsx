"use client";
import { m, AnimatePresence, LazyMotion, domAnimation } from "framer-motion";
import { ChangeEvent, FormEvent, useState } from "react";
import Spinner from "@/components/Spinner";

type Props = {
	duration?: number;
};

export default function Form({ duration = 1 }: Props) {
	const [isSubmitting, setIsSubmitting] = useState(false);
	const [isSubmitted, setIsSubmitted] = useState(false);
	const [data, setData] = useState({
		name: "",
		email: "",
		phone: "",
		message: "",
	});
	const updateData = ({
		target,
	}: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
		setData({
			...data,
			[target.name]: target.value,
		});
	};
	const submit = (e: FormEvent<HTMLFormElement>): void => {
		e.preventDefault();
		setIsSubmitting(true);

		setInterval(() => {
			setIsSubmitting(false);
			setIsSubmitted(true);
		}, 1000);
	};
	return (
		<>
			<LazyMotion features={domAnimation}>
				<AnimatePresence mode="wait">
					{!isSubmitted ? (
						<m.section
							className="grid grid-cols-1 md:grid-cols-[1fr_2fr] place-items-center p-5 mb-5 rounded-lg text-azama-white border border-azama-muted"
							aria-label="section-form"
							initial={{ x: -70, opacity: 0, scale: 0.7 }}
							whileInView={{ x: 0, opacity: 1, scale: 1 }}
							viewport={{once: true}}
							exit={{ opacity: 0 }}
							transition={{ ease: "easeInOut", duration }}
							key={`${isSubmitted}`}
						>
							<div className="mb-5 md:mb-0 md:text-center">
								<h2 className="text-3xl">CONTACT ME!</h2>
								<p>It's free</p>
							</div>
							<form
								onSubmit={submit}
								className="grid gap-5 p-2 place-items-center md:w-full"
								aria-label="contact-form"
							>
								<input
									type="text"
									name="name"
									value={data.name}
									onChange={updateData}
									placeholder="Name *"
									aria-label="name"
									className="bg-azama-base placeholder:text-azama-danger text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5  border focus:border-azama-danger border-azama-muted rounded-xl focus:outline-none"
									required
								/>
								<input
									type="email"
									name="email"
									value={data.email}
									onChange={updateData}
									placeholder="Email *"
									aria-label="email"
									className="bg-azama-base placeholder:text-azama-danger text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5  border border-azama-muted rounded-xl focus:outline-none focus:border-azama-danger"
									required
								/>
								<input
									type="text"
									name="phone"
									value={data.phone}
									onChange={updateData}
									placeholder="Phone"
									aria-label="phone"
									className="bg-azama-base placeholder:text-azama-muted text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5 border border-azama-muted rounded-xl focus:border-azama-primary focus:outline-none"
								/>
								<textarea
									className="bg-azama-base text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-8 border border-azama-muted focus:border-azama-danger rounded-xl placeholder:text-azama-danger focus:outline-none"
									name="message"
									value={data.message}
									onChange={updateData}
									placeholder="Your message goes here... *"
									aria-label="message"
									required
								/>
								<m.button
									className="p-0 m-5 text-5xl rounded-full shadow-md md:w-40 shadow-azama-base bg-azama-base"
									type="submit"
									aria-label="submit"
									whileHover={{ scale: 1.2 }}
									whileTap={{ scale: 0.9 }}
									transition={{ type: "spring", stiffness: 400, damping: 17 }}
								>
									{isSubmitting ? <Spinner /> : "ðŸ”¥"}
								</m.button>
							</form>
						</m.section>
					) : (
						<m.section
							className="grid grid-cols-1 p-5 mb-5 rounded-lg place-items-center text-azama-dark bg-azama-white"
							aria-label="form"
							initial={{ opacity: 0, scale: 0.7 }}
							animate={{ opacity: 1, scale: 1 }}
							transition={{ ease: "easeInOut", duration }}
							key={`${isSubmitted}`}
						>
							<h3 className="text-3xl">
								Thank you!{" "}
								<span className="[text-shadow:_0_3px_0_rgb(35_33_54_/_50%)]">
									ðŸ˜€
								</span>
							</h3>
							<p>
								Your message has been received. Please check your email for
								confirmation.
							</p>
						</m.section>
					)}
				</AnimatePresence>
			</LazyMotion>
		</>
	);
}
