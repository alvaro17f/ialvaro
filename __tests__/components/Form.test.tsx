import { afterEach, beforeEach, describe, expect, it } from "vitest";
import {
	render,
	screen,
	within,
	fireEvent,
	cleanup,
	waitFor,
} from "@testing-library/react";
import Form from "@/app/components/Form";

describe("<Form />", () => {
	beforeEach(() => {
		render(<Form />);
	});
	afterEach(cleanup);

	it("should match the snapshot", () => {
		const { container } = render(<Form />);
		expect(container).toMatchSnapshot();
	});

	it("should render the heading", () => {
		screen.getByRole("heading", { level: 2, name: /contact/i });
	});

	it("should render an exact number of inputs", () => {
		const form = within(screen.getByLabelText("contact-form"));
		expect(form).toBeDefined();
		expect(form.getByRole("textbox", { name: /name/i })).toBeDefined();
		expect(form.getByRole("textbox", { name: /email/i })).toBeDefined();
		expect(form.getByRole("textbox", { name: /phone/i })).toBeDefined();
		expect(form.getByRole("textbox", { name: /message/i })).toBeDefined();
		expect(form.getAllByRole("textbox").length).toBe(4);
	});

	it("should show a loading button on form submission", async () => {
		const form = within(screen.getByLabelText("contact-form"));
		const name = form.getByRole("textbox", { name: /name/i });
		const email = form.getByRole("textbox", { name: /email/i });
		const phone = form.getByRole("textbox", { name: /phone/i });
		const message = form.getByRole("textbox", { name: /message/i });

		fireEvent.change(name, { target: { value: "Alex Turner" } });
		fireEvent.change(email, { target: { value: "arctic@monkeys.com" } });
		fireEvent.change(phone, { target: { value: "66666666" } });
		fireEvent.change(message, {
			target: { value: "Hello World" },
		});

		fireEvent.submit(screen.getByLabelText("submit"));
		await screen.findByText(/loading/i);
	});

	it("should show a 'thank you' message after form submission", async () => {
		const form = within(screen.getByLabelText("contact-form"));
		const name = form.getByRole("textbox", { name: /name/i });
		const email = form.getByRole("textbox", { name: /email/i });
		const phone = form.getByRole("textbox", { name: /phone/i });
		const message = form.getByRole("textbox", { name: /message/i });

		fireEvent.change(name, { target: { value: "Alex Turner" } });
		fireEvent.change(email, { target: { value: "arctic@monkeys.com" } });
		fireEvent.change(phone, { target: { value: "66666666" } });
		fireEvent.change(message, {
			target: { value: "Hello World" },
		});

		fireEvent.submit(screen.getByLabelText("submit"));
		await waitFor(() => {
			screen.getByText(/thank you/i);
		}, {timeout: 2000});
	});
});
