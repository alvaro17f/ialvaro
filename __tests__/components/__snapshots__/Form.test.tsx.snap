// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Form /> > should match the snapshot 1`] = `
<div>
  <section
    aria-label="section-form"
    class="grid grid-cols-1 md:grid-cols-[1fr_2fr] place-items-center p-5 mb-5 rounded-lg text-azama-dark bg-azama-white"
    style="opacity: 0; transform: translateX(-70px) scale(0.7) translateZ(0);"
  >
    <div
      class="mb-5 md:mb-0 md:text-center"
    >
      <h2
        class="text-3xl"
      >
        CONTACT ME!
      </h2>
      <p>
        It's free
      </p>
    </div>
    <form
      aria-label="contact-form"
      class="grid gap-5 p-2 place-items-center md:w-full"
    >
      <input
        aria-label="name"
        class="bg-azama-base placeholder:text-azama-danger text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5 shadow-md rounded-xl shadow-azama-base"
        defaultvalue=""
        name="name"
        placeholder="Name *"
        required=""
        type="text"
      />
      <input
        aria-label="email"
        class="bg-azama-base placeholder:text-azama-danger text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5 shadow-md rounded-xl shadow-azama-base"
        defaultvalue=""
        name="email"
        placeholder="Email *"
        required=""
        type="email"
      />
      <input
        aria-label="phone"
        class="bg-azama-base placeholder:text-azama-muted text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-5 shadow-md rounded-xl shadow-azama-base"
        defaultvalue=""
        name="phone"
        placeholder="Phone"
        type="text"
      />
      <textarea
        aria-label="message"
        class="bg-azama-base text-azama-white max-w-[15rem] md:w-[42rem] md:max-w-none p-8 shadow-md rounded-xl shadow-azama-base placeholder:text-azama-danger"
        name="message"
        placeholder="Your message goes here... *"
        required=""
      />
      <button
        aria-label="submit"
        class="p-4 text-5xl rounded-full shadow-md md:w-40 shadow-azama-base bg-azama-base"
        tabindex="0"
        type="submit"
      >
        ðŸ”¥
      </button>
    </form>
  </section>
</div>
`;
